src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -0,0 +1,4787 @@
+
+	return idx;
+}
+
+static int
+find_next_available_dirfd(struct dir_obj *obj)
+{
+	int i, idx = -1, rc;
+	struct dir_obj *new_obj;
+	bool allocated = false;
+
+	if (obj == NULL) {
+		D_ALLOC(new_obj, sizeof(struct dir_obj));
+		if (new_obj == NULL) {
+			errno = ENOMEM;
+			return (-1);
+		}
+		new_obj->dir       = NULL;
+		new_obj->ref_count = 0;
+		allocated          = true;
+	} else {
+		new_obj            = obj;
+	}
+
+	rc = pthread_mutex_lock(&lock_dirfd);
+	D_ASSERT(rc == 0);
+	if (next_free_dirfd < 0) {
+		rc = pthread_mutex_unlock(&lock_dirfd);
+		D_ASSERT(rc == 0);
+		if (allocated)
+			D_FREE(new_obj);
+		return next_free_dirfd;
+	}
+	idx = next_free_dirfd;
+	if (idx >= 0) {
+		new_obj->ref_count++;
+		dir_list[idx] = new_obj;
+	}
+	if (next_free_dirfd > last_dirfd)
+		last_dirfd = next_free_dirfd;
+
+	next_free_dirfd = -1;
+
+	for (i = idx + 1; i < MAX_OPENED_DIR; i++) {
+		if (dir_list[i] == NULL) {
+			/* available, then update next_free_dirfd */
+			next_free_dirfd = i;
+			break;
+		}
+	}
+	if (next_free_dirfd < 0)
+		printf("WARNING> All space for dir_list are used.\n");
+
+	num_dirfd++;
+	rc = pthread_mutex_unlock(&lock_dirfd);
+	D_ASSERT(rc == 0);
+
+	if (idx < 0 && allocated)
+		D_FREE(new_obj);
+
+	return idx;
+}
+
+static int
+find_next_available_map(void)
+{
+	int i, idx = -1, rc;
+
+	rc = pthread_mutex_lock(&lock_mmap);
+	D_ASSERT(rc == 0);
+	if (next_free_map < 0) {
+		rc = pthread_mutex_unlock(&lock_mmap);
+		D_ASSERT(rc == 0);
+		return next_free_map;
+	}
+	idx = next_free_map;
+	if (next_free_map > last_map)
+		last_map = next_free_map;
+	next_free_map = -1;
+
+	for (i = idx + 1; i < MAX_MMAP_BLOCK; i++) {
+		if (mmap_list[i].addr == NULL) {
+			/* available, then update next_free_map */
+			next_free_map = i;
+			break;
+		}
+	}
+	if (next_free_map < 0)
+		printf("WARNING> All space for mmap_list are used.\n");
+
+	num_map++;
+	rc = pthread_mutex_unlock(&lock_mmap);
+	D_ASSERT(rc == 0);
+
+	return idx;
+}
+
+/* May need to support duplicated fd as duplicated dirfd too. */
+static void
+free_fd(int idx)
+{
+	int              i, rc;
+	struct file_obj *saved_obj = NULL;
+
+	rc = pthread_mutex_lock(&lock_fd);
+	D_ASSERT(rc == 0);
+	if (file_list[idx]->idx_mmap >= 0 && file_list[idx]->idx_mmap < MAX_MMAP_BLOCK) {
+		/* file_list[idx].idx_mmap >= MAX_MMAP_BLOCK means fd needs closed after munmap()*/
+		file_list[idx]->idx_mmap += MAX_MMAP_BLOCK;
+		rc = pthread_mutex_unlock(&lock_fd);
+		D_ASSERT(rc == 0);
+		return;
+	}
+
+	file_list[idx]->ref_count--;
+	if (file_list[idx]->ref_count == 0)
+		saved_obj = file_list[idx];
+	file_list[idx] = NULL;
+
+	if (idx < next_free_fd)
+		next_free_fd = idx;
+
+	if (idx == last_fd) {
+		for (i = idx - 1; i >= 0; i--) {
+			if (file_list[i]) {
+				last_fd = i;
+				break;
+			}
+		}
+	}
+
+	num_fd--;
+	rc = pthread_mutex_unlock(&lock_fd);
+	D_ASSERT(rc == 0);
+
+	if (saved_obj) {
+		rc = dfs_release(saved_obj->file);
+		memset(saved_obj, 0, sizeof(struct file_obj));
```

Comment:
This looks like debug hacks left in by mistake.

Reply:
I can add a comment there. This was left there intended. In case of duplicated fd exists, multiple fd could point to same struct file_obj. struct file_obj is supposed to be freed only when reference count reaches zero. With zeroing out the memory, we could observe issues immediately if something is wrong (e.g., one open fd is still points to this struct). We can remove this later after we make sure the code about adding and decreasing the reference to the struct.

Reply:
We will remove "memset(saved_obj, 0, sizeof(struct file_obj));" later in another PR. Thank you!